src/aspire/utils/blk_diag_matrix.py

```diff
@@ -0,0 +1,700 @@
+"""
+Define a BlkDiagMatrix module which implements operations for
+block diagonal matrices as used by ASPIRE.
+"""
+
+import numpy as np
+
+from numpy.linalg import norm
+from numpy.linalg import solve
+from scipy.linalg import block_diag
+from scipy.special import jv
+
+from aspire.utils.cell import Cell2D
+from aspire.basis.ffb_2d import FFBBasis2D
+from aspire.basis.basis_utils import lgwt
+
+
+class BlkDiagMatrix:
+    """
+    Define a BlkDiagMatrix class which implements operations for
+    block diagonal matrices as used by ASPIRE.
+    """
+
+    # All instances of this class should have priority over ndarray ops
+    #   because we implement them here ourselves.
+    # This is a more np current implementation of __array_priority__
+    #   operator precedence schedule.
+    # Mainly this effects rmul, radd, rsub eg:
+    #   blk_y = scalar_a + ( scalar_b * blk_x)
+    __array_ufunc__ = None
+
+    def __init__(self, nblocks, partition=None, dtype=np.float64):
+        """
+        Instantiate a BlkDiagMatrix.
+
+        :param nblocks: Number of diagonal matrix blocks
+        :param partition: Optionally supply the matrix block partition
+         in the form of a `nblock`-element list storing all shapes of
+         diagonal matrix blocks, where `partition[i]` corresponds to
+         the shape (number of rows and columns) of the `i` matrix block.
+        :return BlkDiagMatrix instance
+
+        """
+
+        self.nblocks = nblocks
+        self.dtype = np.dtype(dtype)
+        self.data = [[]] * nblocks
+        self._cached_blk_sizes = None
+        if partition:
+            self._cached_blk_sizes = np.array(partition)
```

Comment:
It is not totally necessary, but caching the partition structure on first touch as a np array is more efficient than repeated looping over python lists for every use of an arithmetic method on the BlkDiagMatrix instances.  I would prefer to leave it, but if you have strong feelings about it, it can certainly be removed. We would simply go back to deep looping over the list.

Reply:
Ah right. That makes sense. No this was just for my information rather than anything else.